### DISCUSSION TITLE
Sync For Specific User

### DISCUSSION THREAD

**Discussion by Diledio** [Mimetis/Dotmim.Sync#1051] [Q&A]
Hello guys!
I have one quastion that i am not solving, 
i am developing an app in xamarin working with sql lite and get also a sql server to get and insert data using this dotmim.sync,
For example this app is using by 10 users and each user now get and insert data without any problem but what i want is to sync data for each specific user, in this xamarin app user login and after login is the sync button to get and insert data, but i want to sync data for this user who login because for now it sync all data, i add filter but still nothing worked, hope i can get a answer thank you 
Here is the code behind in the button clicked event after he login 

private async void  Button_Clicked_1(object sender, EventArgs e)
        {

            try
            {
                int userId = Global.loginId;
                
                SQLitePCL.Batteries_V2.Init();

                var serverOrchestrator = new WebRemoteOrchestrator("http://10.150.2.101:45455/api/sync");

                var sqliteFilename = "WHMAGNA.db";
                string documentsDirectoryPath = System.Environment.GetFolderPath(System.Environment.SpecialFolder.Personal);
                var path = Path.Combine(documentsDirectoryPath, sqliteFilename);

                var clientProvider = new SqliteSyncProvider(path);
                var options = new SyncOptions
                {
                    BatchDirectory = Path.Combine(SyncOptions.GetDefaultUserBatchDirectory(), "client")
                };
               
                
                var agent = new SyncAgent(clientProvider, serverOrchestrator, options);
                var parameters = new SyncParameters
                {

                 { "loginId" ,  userId }

                };
                var s1 = await agent.SynchronizeAsync("testi103", parameters);
                
                lblResult.Text = "Sinkronizimi mbaroi";
                lblResult.FontAttributes = FontAttributes.Bold;
                lblResult.FontSize = 35;
                lblResult.TextColor = Color.LightSeaGreen;
                string syncResult = lblResult.Text;

                lblResult.Text = string.Empty;
                await DisplayAlert("Sync Completed", syncResult, "OK");

                
            }
            catch (Exception ex)
            {
                lblResult.Text = "Error: " + ex.Message;
                lblResult.FontAttributes = FontAttributes.Bold;
                lblResult.FontSize = 35;
                lblResult.TextColor = Color.OrangeRed;
                string syncResult = lblResult.Text;

                // Reset the label text
                lblResult.Text = string.Empty;
                await DisplayAlert("Sync Error", syncResult, "OK");
            }

        }

Here int userId = Global.loginId; is the current user who logged in and here the parameters var parameters = new SyncParameters
                {

                 { "loginId" ,  userId }

                };

here is the code in server side 

 public void ConfigureServices(IServiceCollection services)
        {
            try
            {
                
                services.AddControllers();

                //services.AddDistributedMemoryCache();
                //services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));

                services.AddDistributedMemoryCache();
                services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));
                //services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));
               
                var connectionString = Configuration.GetSection("ConnectionStrings")["DefaultConnection"];
                var options = new SyncOptions
                {
                    BatchDirectory = Path.Combine(SyncOptions.GetDefaultUserBatchDirectory(), "server")
                };
                var setup = new SyncSetup("aUsers", "D_Ngarkese_Pirunist", "cMovs");
                
               

                var filter = new SetupFilter("aUsers");
                
                filter.AddParameter("loginId", System.Data.DbType.String);
                filter.AddWhere("loginId", "aUsers", "loginId");
                setup.Filters.Add(filter);

                services.AddSyncServer<SqlSyncProvider>(connectionString, "testi103", setup, options);
                
                services.AddSwaggerGen();
            }
            catch (Exception ex)
            {
                string sd = ex.Message;
            }
            
        }

**Comment by AlexHolly** [comment]
@Diledio it would help if you can provide the version you are using.

I also use an Id to filter the data maybe my issue is related.
I tried "Dotmim.Sync.MariaDB" Version="1.0.0-beta-1517" and
          "Dotmim.Sync.MySql" Version="1.0.0-beta-1517"

With version 1.0.0-beta-1517 it seem there is never an upload happening in case of `SyncDirection.Bidirectional`. The API call is happening but no data is changing on the server.

Also tried with version 0.9.8. It looks like 0.9.8 works a bit better in case of `SyncDirection.Bidirectional` but to make sync work I have to make changes on client, then on the server and to trigger the sync again on the client to make it work as expected which is strange because I expect a sync everytime I make a change on the client. The API call is happening but no data is changing on the server.

1. It would be good to know which package should be used in case of MariaDb. Dotmim.Sync.MySql or Dotmim.Sync.MariaDB so far I don't see a difference. Is one of them more stable?

2. It is the first time I use this framework maybe I expect too much from SyncDirection.Bidirectional? It is still a bit strange that version 0.9.8 and 1.0.0-beta-1517 behave differently. Are there bigger changes between the versions in this case? Is it a known issue in version 0.9.8 to make the sync work properly, changes have to be made first on both sides? @Mimetis

I will try to provide a minimal example as soon as I knows which version makes most sense right now.

**Reply by Mimetis** [reply]
The last version (even if it's a still a beta version) `1.0` is the one you may want to use.

I don't have a lot of people using the `MariaDB `/ `MySQL `providers, that's why I don't have a lot of feedbacks
I tried, the most I can do, to have the same tests coverages for all providers.
But I may still have some issues, tbh.

If you see something that is not working properly, do not hesitate to raise a bug.
And if you can provide a sample, or at least a `MySQL` script to use and a bit of code, to reproduce the issue, that will help a lot.

Looking at closely to your message:

> With version 1.0.0-beta-1517 it seem there is never an upload happening in case of SyncDirection.Bidirectional. The API call is happening but no data is changing on the server.

That should not happens this way. If you are making a change locally on the client (and that row is part of the filter) the change should be reflected on the server.

> Also tried with version 0.9.8. It looks like 0.9.8 works a bit better in case of SyncDirection.Bidirectional 

That should not happens, at least `v1.0` should do same (or better) than `v0.9.8`, at least.

> It would be good to know which package should be used in case of MariaDb. Dotmim.Sync.MySql or Dotmim.Sync.MariaDB so far I don't see a difference. Is one of them more stable?

The two providers are more or less the same. Only namespaces are changing for consistency

> It is the first time I use this framework maybe I expect too much from SyncDirection.Bidirectional? It is still a bit strange that version 0.9.8 and 1.0.0-beta-1517 behave differently. 

Should not

> Are there bigger changes between the versions in this case? Is it a known issue in version 0.9.8 to make the sync work properly, changes have to be made first on both **sides**

Not that much honestly ...


For each issue, if you can provide a sample repro script, that would help (and if you can use the AdventureWorks database, using one of the scripts from the root repository, it's even better)

Let me know

**Reply by AlexHolly** [reply]
In my use case I export the MariaDb first to Sqlite which is working well but in case of "AdventureWorks" Database it fails. So I am unable to reproduce my actual issue right now. Do you have an idea?

![image](https://github.com/Mimetis/Dotmim.Sync/assets/4741886/8f3ccaec-f811-46e8-9e41-05c8b89e982d)

Here is the "minimal" example solution
[MariaDbWebSyncExample.zip](https://github.com/Mimetis/Dotmim.Sync/files/13024728/MariaDbWebSyncExample.zip)

**Reply by AlexHolly** [reply]
@Mimetis Here another version which is very minimal it shows the issue very well.

Create Database "adventureworksmin"
Create Table
```
CREATE TABLE `salesorderdetail` (
	`SalesOrderDetailID` INT(11) NOT NULL AUTO_INCREMENT,
	`ProductID` INT(11) NOT NULL,
	`ModifiedDate` DATETIME NOT NULL,
	PRIMARY KEY (`SalesOrderDetailID`) USING BTREE
)
ENGINE=InnoDB
;
```

1. Version 1.0.0-beta-1517
1.1 Export will also include the tables ["*_tracking", "scope_info", "scope_info_client"] -> This is not configured in the setup but happens anyway.
1.2 After Exporting and Cleanup of (1.1) you can add new entries with the SqliteClient prohect and press "Sync Database with Server".

2. Version 0.9.8
2.1 Here Sync works as expected (1.2)
2.2 Changes on the MariaDbServer are not tracked, tables like "scope_info", ... are not created until first Sync is excuted by client.

[MariaDbWebSyncExampleMin.zip](https://github.com/Mimetis/Dotmim.Sync/files/13040050/MariaDbWebSyncExampleMin.zip)

**Reply by AlexHolly** [reply]
@Mimetis Any updates on the issue? Should I open a bug for it?

**Reply by Mimetis** [reply]
I'm sorry, I don't download any zip file.
If you can provide a Github repo with console sample reproducing the error, it will help
I will look into it as soon as I have free time

**Reply by AlexHolly** [reply]
https://github.com/AlexHolly/reproduce-sync-issue-mysql
